tf_encrypted/protocol/securenn/securenn.py

```diff
@@ -458,50 +443,67 @@ def _bits_public(prot, x: PondPublicTensor,
         x_on_0, x_on_1 = x.unwrapped
 
         with tf.device(prot.server_0.device_name):
-            bits_on_0 = x_on_0.to_bits(factory)
+            bits_on_0 = x_on_0.bits(factory)
 
         with tf.device(prot.server_1.device_name):
-            bits_on_1 = x_on_1.to_bits(factory)
+            bits_on_1 = x_on_1.bits(factory)
 
         return PondPublicTensor(prot, bits_on_0, bits_on_1, False)
 
 
-def _lsb_private(prot, y: PondPrivateTensor):
+def _lsb_public(prot, x: PondPublicTensor):
+    # TODO[Morten]
+    # we chould call through and ask the underlying dtype for its lsb instead as there might
+    # be more efficient ways of getting it for some types (ie without getting all bits)
+    x_bits = prot.bits(x)
+    x_lsb = x_bits[..., 0]
+    return x_lsb
+
+
+def _lsb_private(prot, x: PondPrivateTensor):
+
+    # TODO[Morten] in the refactor these could be type parameters
+    odd_dtype = x.backing_dtype
+    out_dtype = prot.tensor_factory
+    prime_dtype = prot.prime_factory
+
+    assert odd_dtype.modulus % 2 == 1
+    assert x.backing_dtype == odd_dtype  # needed for security because of `r` masking
 
     with tf.name_scope('lsb'):
 
         with tf.name_scope('lsb_mask'):
 
             with tf.device(prot.server_2.device_name):
-                x_raw = y.backing_dtype.sample_uniform(y.shape)
-                xbits_raw = x_raw.to_bits(factory=prot.prime_factory)
-                xlsb_raw = xbits_raw[..., 0]
+                r0_raw = odd_dtype.sample_uniform(x.shape)
```

Comment:
Curious why not just `_share`?

Reply:
good catch, believe we could change it now if that's better.

motivation was that `_share` works differently: it draws one sample and subtracts that from the value. here we are essentially sharing a random value, so doing it this way gives exactly the same; however, no subtraction is involved. in the end I implemented subtraction for odd tensors as well so `_share` could be used, yet this is likely how we'll share random tensors after introducing seeding.